compiler/src/dotty/tools/dotc/transform/SuperPhase.scala

```diff
@@ -0,0 +1,625 @@
+
+  /** Does `phase` contain a redefinition of method `name`?
+   *  (which is a method of MiniPhase)
+   */
+  private def defines(phase: MiniPhase, name: String) = {
+    def hasRedefinedMethod(cls: Class[_]): Boolean =
+      if (cls.eq(classOf[MiniPhase])) false
+      else {
+        var clsMethods = clsMethodsCache.get(cls)
+        if (clsMethods eq null) {
+          clsMethods = cls.getDeclaredMethods
+          clsMethodsCache.put(cls, clsMethods)
+        }
+        clsMethods.exists(_.getName == name) ||
+        hasRedefinedMethod(cls.getSuperclass)
+      }
+    hasRedefinedMethod(phase.getClass)
+  }
+
+  /** A transformer array is an array of node or context transformers. It has
+   *  one more element than there are miniphases. It is constructed as follows:
+   *
+   *  - the last element is always `last`, which is the identity transformer
+   *    that returns the context or tree unchanged.
+   *  - The element corresponding to phase P is `elemFn(P)` if `P` defines a
+   *    method with name `methName`, or it the same as the following element
+   *    if `P` does not define `methName`.
+   */
+  private def newTransformerArray[T, R](
+      methName: String, elemFn: IndexedTransformer[T, R], last: Transformer[T, R]) = {
+    val trans = new Array[Transformer[T, R]](miniPhases.length + 1)
+    trans(miniPhases.length) = last
+    for (idx <- miniPhases.length - 1 to 0 by -1) {
+      val subPhase = miniPhases(idx)
+      trans(idx) = if (defines(subPhase, methName)) elemFn(subPhase) else trans(idx + 1)
+    }
+    trans
+  }
+
+  private def newContextTransformerArray[T](suffix: String, elemFn: IndexedTransformer[T, Context]) =
+    newTransformerArray[T, Context]("prepareFor" + suffix, elemFn, idContextTransformer.asInstanceOf[Transformer[T, Context]])
+
+  private def newNodeTransformerArray[T, R](suffix: String, elemFn: IndexedTransformer[T, R]) =
+    newTransformerArray[T, R]("transform" + suffix, elemFn, idNodeTransformer.asInstanceOf[Transformer[T, R]])
+
+  private val statsContextTransformer = newContextTransformerArray("Stats", prepForStats)
+  private val statsNodeTransformer    = newNodeTransformerArray("Stats", transStats)
+  private val unitContextTransformer  = newContextTransformerArray("Unit", prepForUnit)
+  private val unitNodeTransformer     = newNodeTransformerArray("Unit", transUnit)
+  private val otherContextTransformer = newContextTransformerArray("Other", prepForOther)
+  private val otherNodeTransformer    = newNodeTransformerArray("Other", transOther)
+
+  private val contextTransformer: Array[Array[Transformer[Tree, Context]]] =
+    Array.fill(Tag.NumTags)(otherContextTransformer)
+  private val nodeTransformer: Array[Array[Transformer[Tree, Tree]]] =
+    Array.fill(Tag.NumTags)(otherNodeTransformer)
+
+  private def init(name: String, tag: TreeTag, ctf: IndexedTransformer[_, Context], ntf: IndexedTransformer[_, Tree]): Unit = {
```

Comment:
I think `init` is too generic name and also counterintuitive given that it is called more than once.